distributed/core.py

```diff
@@ -109,6 +109,106 @@ def _expects_comm(func: Callable) -> bool:
+    closed: bool
+
+    def __init__(self):
+        self.closed = False
+        self._ongoing_tasks = set()
+
+    def call_soon(
+        self, afunc: Callable[..., Coroutine], *args, **kwargs
+    ) -> asyncio.Task | None:
+        """Schedule a coroutine function to be executed as an `asyncio.Task`.
+
+        The coroutine function `afunc` is scheduled with `args` arguments and `kwargs` keyword arguments
+        as an `asyncio.Task`.
+
+        Parameters
+        ----------
+        afunc
+            Coroutine function to schedule.
+        *args
+            Arguments to be passed to `afunc`.
+        **kwargs
+            Keyword arguments to be passed to `afunc`
+
+        Returns
+        -------
+        asyncio.Task | None
+            The scheduled Task object, or None if the group is closed.
+        """
+        if self.closed:
+            return None
+
+        task = asyncio.create_task(afunc(*args, **kwargs))
+        self._ongoing_tasks.add(task)
+        task.add_done_callback(self._ongoing_tasks.remove)
+        return task
+
+    def call_later(
+        self, delay: int, afunc: Callable[..., Coroutine], *args, **kwargs
+    ) -> asyncio.Task | None:
+        """Schedule a coroutine function to be executed after `delay` seconds as an `asyncio.Task`.
+
+        The coroutine function `afunc` is scheduled with `args` arguments and `kwargs` keyword arguments
+        as an `asyncio.Task` that is executed after `delay` seconds.
+
+        Parameters
+        ----------
+        delay
+            Delay in seconds.
+        afunc
+            Coroutine function to schedule.
+        *args
+            Arguments to be passed to `afunc`.
+        **kwargs
+            Keyword arguments to be passed to `afunc`
+
+        Returns
+        -------
+        asyncio.Task | None
+            The scheduled Task object, or None if the group is closed.
+        """
+        return self.call_soon(delayed(afunc, delay), *args, **kwargs)
+
+    @property
+    def _cancellable_tasks(self):
+        return (t for t in self._ongoing_tasks if t is not asyncio.current_task())
+
+    def close(self):
+        """Closes the task group so that no new tasks can be scheduled.
+
+        Existing tasks continue to run.
+        """
+        self.closed = True
+
+    async def stop(self, timeout=1):
+        """Close the group and stop all currently running tasks.
+
+        Closes the task group and waits `timeout` seconds for all tasks to gracefully finish.
+        After the timeout, all remaining tasks are cancelled.
+        """
+        self.close()
+        try:
+            gather = asyncio.gather(
+                *self._cancellable_tasks,
+                return_exceptions=True,
+            )
+            await asyncio.wait_for(
+                gather,
+                timeout,
+            )
+        except asyncio.TimeoutError:
+            try:
+                await gather
+            except asyncio.CancelledError:
```

Comment:
because `return_exceptions=True` was passed to gather that would mean this CancelledError was raised because someone cancelled the call to TaskGroup.stop() ?

Reply:
seems related https://github.com/python/cpython/issues/73618

for me wrapping the call to gather in an async def also swallows the `_GatheringFuture exception was never retrieved`:

```python
import asyncio

async def foo():
    await asyncio.sleep(10)

async def bar():
    async def async_fn():
        return await asyncio.gather(foo())

    try:
        await asyncio.wait_for(async_fn(), timeout=1)
    except asyncio.TimeoutError:  # wait_for timed out after 1 second
        pass

asyncio.run(bar())
```

Reply:
Yes, I came across that issue yesterday as well. Let's go with the `async def` version then, it looks less convoluted. I'll link to this discussion in a comment.

Reply:
wrapping the `_GatheringFuture` in task also causes `_GatheringFuture.result` to get called: https://github.com/python/cpython/blob/372afb7a9b1fff2fcc724f999744b6126f0f6e07/Lib/asyncio/tasks.py#L304

Reply:
TIL: Wrapping the `_GatheringFuture` in `asyncio.create_task()` does not work after all, it throws a `TypeError: a coroutine was expected, got <_GatheringFuture pending>`.

```python3
import asyncio

async def foo():
    await asyncio.sleep(10)

async def bar():
    try:
        task = asyncio.create_task(asyncio.gather(foo()))
        await asyncio.wait_for(task, timeout=1)
    except asyncio.TimeoutError:  # wait_for timed out after 1 second
        try:
            await task
        except asyncio.CancelledError:
            pass

asyncio.run(bar())
```
```
Traceback (most recent call last):
  File "/Users/hendrikmakait/projects/dask/distributed/distributed/script.py", line 19, in <module>
    asyncio.run(bar())
  File "/opt/homebrew/Caskroom/mambaforge/base/envs/dask-distributed/lib/python3.9/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/opt/homebrew/Caskroom/mambaforge/base/envs/dask-distributed/lib/python3.9/asyncio/base_events.py", line 647, in run_until_complete
    return future.result()
  File "/Users/hendrikmakait/projects/dask/distributed/script.py", line 10, in bar
    task = asyncio.create_task(asyncio.gather(foo()))
  File "/opt/homebrew/Caskroom/mambaforge/base/envs/dask-distributed/lib/python3.9/asyncio/tasks.py", line 361, in create_task
    task = loop.create_task(coro)
  File "/opt/homebrew/Caskroom/mambaforge/base/envs/dask-distributed/lib/python3.9/asyncio/base_events.py", line 438, in create_task
    task = tasks.Task(coro, loop=self, name=name)
TypeError: a coroutine was expected, got <_GatheringFuture pending>
```